Add shouldGetAll to `.identify()`, correct documentation

vladh opened the pull request:
Hi!

I, as in #2, also needed to get all of the names for a chord based on a set of notes. I decided to add a `shouldGetAll` option, which simply returns all of the inversions. This presented a small problem for `.identify(notes...)` which takes the notes from the argument list, as there is no room for the `shouldGetAll` argument in a way that makes sense. So I only implemented it for `.identifyArray(notes)`, and changed `.identify(notes...)` to be based off of it. I've also changed the documentation accordingly.

I don't know if you think this is the best approach or not. I'm welcome to suggestions, it's just the best and most compatible way to achieve something I needed. An alternative without the extra argument would be creating a `.identifyArrayAll()` method or similar.

One more thing — regardless of this change, the documentation incorrectly stated that `.identify()` returns a `Chord`, whereas it only returns a chord symbol. I've also corrected this in the edited documentation.

Looking forward to hearing what you think!

vladh commented:
An improvement to this would be to only return unique symbols. Here's an example.

```
> s11.chord.identifyArray(['F', 'C', 'F', 'A', 'D', 'E'], true)
[ { symbol: 'FM13', reasonable: true },
  { symbol: 'C6/9add11/F', reasonable: false },
  { symbol: 'FM13', reasonable: true },
  { symbol: 'Amadd11/F', reasonable: false },
  { symbol: 'Dm9/F', reasonable: true },
  { symbol: 'Esus47b9/F', reasonable: false } ]
```

I didn't implement this as I'm not sure if the reasoning is sound, but can definitely do it.

jsrmath commented:
I think I'd prefer this to be a separate function called `getPossibleChordNames()` and then a corresponding `getPossibleChordNamesFromArray()`.  I agree that it should only return unique symbols, which you can do with `_.uniq()`.  I'm happy to make this change myself, or if you'd prefer to do it, just abstract out the common logic from `identify` into a function called `getPossibleChords()`.

I also think `getPossibleChordNames()` should just return a list of strings and ignore the `reasonable` parameter, but let me know if you feel strongly otherwise.

And thanks for catching that error in the docs.  I'll just change that now since it's a quick fix.

vladh commented:
Hm, yes, I think that's a better idea. I've updated the code with these two new functions, with the original code being refactored into a function used both by `getPossibleChordNames()` and `identify()`. Only unique matches are returned now.

Regarding the `reasonable` parameter, I agree, but have sorted the list of chords such that reasonable names come first, as I found it a useful feature.

I've also updated the documentation, I hope it's clear enough.

jsrmath commented on lib/chord.js:
```diff
@@ -321,6 +321,30 @@ var identify = function () {
 
   var results = _.map(inversions, tryInversion);
```
I'd prefer if you combined all the variables that come next using `_.chain()` and then just return here.

jsrmath commented on lib/chord.js:
```diff
@@ -321,6 +321,30 @@ var identify = function () {
+var getPossibleChordNames = function () {
+  return getPossibleChords
+    .apply(this, arguments)
+    .map(function (n) { return n.symbol });
```
Put the return on a new line.  I don't generally use this style in Sharp11.

jsrmath commented on docs/chord.md:
```diff
@@ -13,6 +13,12 @@ Returns a [Chord](#chord-object) object given an argument list of [Note](note.md
 ### <a name="module-identify-array"></a> identifyArray `.identifyArray([notes])`
 Invokes `.identify()` with the contents of `[notes]` as its arguments.
 
+### <a name="module-identify"></a> getPossibleChordNames `.getPossibleChordNames(notes...)`
```
Change the name property of the <a> tag here

jsrmath commented on docs/chord.md:
```diff
@@ -13,6 +13,12 @@ Returns a [Chord](#chord-object) object given an argument list of [Note](note.md
+### <a name="module-identify"></a> getPossibleChordNames `.getPossibleChordNames(notes...)`
+Returns a list of chord names given an argument list of [Note](note.md#note-object) objects or strings, ignoring octave numbers. Similar to `.identify(notes...)`, but returns all possible matches. Matches are sorted by how "reasonable" they are, with reasonable ones coming first in the list.
+
+### <a name="module-identify-array"></a> getPossibleChordNamesFromArray `.getPossibleChordNamesFromArray([notes])`
```
And here

jsrmath commented:
Oh sorry, one more thing.  Would you mind terribly writing test cases for these?

vladh commented:
Should all be done. Please look over it real quick, just to make sure all the correct changes are there after the rebase.

I hope the tests are good enough, I didn't go crazy with them, just wrote what I thought was sufficient.

Hope everything is in order! 😄
